File: android/clawchat-kit/src/main/kotlin/com/clawchat/kit/protocol/ConnectionMessages.kt
package com.clawchat.kit.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID

// MARK: - App Platform

@Serializable
enum class AppPlatform {
    @SerialName("ios") IOS,
    @SerialName("android") ANDROID,
    @SerialName("web") WEB,
    @SerialName("cli") CLI,
}

// MARK: - App Pair

@Serializable
data class AppPair(
    override val type: String = "app.pair",
    override val id: String = UUID.randomUUID().toString(),
    override val ts: Long = System.currentTimeMillis(),
    val pairingCode: String,
    val deviceName: String,
    val platform: AppPlatform = AppPlatform.ANDROID,
    val protocolVersion: String = "0.1",
) : BaseMessage

// MARK: - App Paired

@Serializable
data class AppPaired(
    override val type: String,
    override val id: String,
    override val ts: Long,
    val deviceToken: String,
    val gatewayId: String,
) : BaseMessage

// MARK: - App Pair Error

@Serializable
enum class PairErrorReason {
    @SerialName("invalid_code") INVALID_CODE,
    @SerialName("code_expired") CODE_EXPIRED,
    @SerialName("expired") EXPIRED,
    @SerialName("gateway_offline") GATEWAY_OFFLINE,
}

@Serializable
data class AppPairError(
    override val type: String,
    override val id: String,
    override val ts: Long,
    val error: PairErrorReason,
    val message: String,
) : BaseMessage

// MARK: - App Connect

@Serializable
data class AppConnect(
    override val type: String = "app.connect",
    override val id: String = UUID.randomUUID().toString(),
    override val ts: Long = System.currentTimeMillis(),
    val deviceToken: String,
    val protocolVersion: String = "0.1",
) : BaseMessage

// MARK: - App Connected

@Serializable
data class AppConnected(
    override val type: String,
    override val id: String,
    override val ts: Long,
    val gatewayId: String,
    val gatewayOnline: Boolean? = null,
    val newDeviceToken: String? = null,
) : BaseMessage

File: ios/ClawChatKit/Sources/ClawChatKit/Protocol/ConnectionMessages.swift
import Foundation

// MARK: - App Platform

public enum AppPlatform: String, Codable, Sendable {
    case ios, android, web, cli
}

// MARK: - App Pair

public struct AppPair: BaseMessage, Sendable {
    public let type: String = "app.pair"
    public let id: String
    public let ts: Int64
    public let pairingCode: String
    public let deviceName: String
    public let platform: AppPlatform
    public let protocolVersion: String

    public init(pairingCode: String, deviceName: String, platform: AppPlatform = .ios, protocolVersion: String = "0.1") {
        self.id = UUID().uuidString
        self.ts = Int64(Date().timeIntervalSince1970 * 1000)
        self.pairingCode = pairingCode
        self.deviceName = deviceName
        self.platform = platform
        self.protocolVersion = protocolVersion
    }
}

// MARK: - App Paired

public struct AppPaired: BaseMessage, Sendable {
    public let type: String
    public let id: String
    public let ts: Int64
    public let deviceToken: String
    public let gatewayId: String
    public let agents: [String]?
}

// MARK: - App Pair Error

public enum PairErrorReason: String, Codable, Sendable {
    case invalidCode = "invalid_code"
    case codeExpired = "code_expired"
    case expired
    case gatewayOffline = "gateway_offline"
}

public struct AppPairError: BaseMessage, Sendable {
    public let type: String
    public let id: String
    public let ts: Int64
    public let error: PairErrorReason
    public let message: String
}

// MARK: - App Connect

public struct AppConnect: BaseMessage, Sendable {
    public let type: String = "app.connect"
    public let id: String
    public let ts: Int64
    public let deviceToken: String
    public let protocolVersion: String

    public init(deviceToken: String, protocolVersion: String = "0.1") {
        self.id = UUID().uuidString
        self.ts = Int64(Date().timeIntervalSince1970 * 1000)
        self.deviceToken = deviceToken
        self.protocolVersion = protocolVersion
    }
}

// MARK: - App Connected

public struct AppConnected: BaseMessage, Sendable {
    public let type: String
    public let id: String
    public let ts: Int64
    public let gatewayId: String
    public let gatewayOnline: Bool?
    public let agents: [String]?
    public let newDeviceToken: String?
}
